### PR TITLE
#37435 prevent multiple loading of deployment config

### LINKED ISSUE #37435: [Performance] DeploymentConfig - reloadData / flattenParams
### Preconditions and environment

- Magento version 2.4.6

### Steps to reproduce

1. Profile a frontend product page loading.

### Expected result

DeploymentConfig is only loaded once

### Actual result

Multiple call to `DeploymentConfig::reloadData`

### Additional information

Capture d’écran 2023-04-26 à 11 47 12

I'm not sure to fully understand how DeploymentConfig class is working, so maybe I'm wrong and those 266 calls to `DeploymentConfig::reloadData()` are needed. In that case, we should try to optimize flattenParams which is the time consuming method here.

But if I understand it correctly, there is no reason to load that config multiple times for a single request (even if some config path is null).
In that case, I'm not sure what the best fix is, but I'm sure fixing it would improve significantly frontend pages.

### Release note

_No response_

### Triage and priority

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [X] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by Nuranto**
see #37435

### Fixed Issues (if relevant)

1. Fixes magento/magento2#37435

### Manual testing scenarios (*)

1. ...
2. ...

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by Nuranto** [comment]
@magento run all tests

**Comment by Nuranto** [comment]
@ihor-sviziev 

Before : 

https://user-images.githubusercontent.com/1633463/234538809-f828855b-59c4-4ec7-a866-f2b0a0cdb464.png

After : 
Capture d’écran 2023-04-26 à 14 41 53

**Comment by ihor-sviziev** [comment]
@Nuranto just checked on my installation, we don't have such an issue for some reason. So would be great to add some steps to reproduce of this issue.
Also, there are some test failures.

**Comment by Nuranto** [comment]
In my environment, those two config variables were not set anywhere : 
* static_content_on_demand_in_production
* force_html_minification

And we are using html minification, and were running on production mode.

That was causing the most of duplicated calls to `reloadData()`. You can see a part of the trace on the "before" screenshot.

So I guess the steps to reproduces are : 
1. Production mode
2. enable html minification
3. make sure `static_content_on_demand_in_production` and `force_html_minification` are null

**Comment by Nuranto** [comment]
@magento run all tests

File: lib/internal/Magento/Framework/App/DeploymentConfig.php [hunk 1]
```diff
@@ -170,8 +158,12 @@ private function getEnvOverride() : array
      * @throws FileSystemException
      * @throws RuntimeException
      */
-    private function reloadData(): void
+    private function loadData(): void
     {
+        if (!empty($this->data) && $this->isAvailable()) {
```
Comment (ihor-sviziev) [inline lib/internal/Magento/Framework/App/DeploymentConfig.php]: maybe we can use comparison with null value, it mean it won't trigger not needed load when "data" is empty array?
```suggestion
        if ($this->data! == null && $this->isAvailable()) {
```

**Comment by Nuranto** [comment]
@magento run Unit Tests

File: lib/internal/Magento/Framework/App/DeploymentConfig.php (hunk 1 above)
Comment (Nuranto) [reply lib/internal/Magento/Framework/App/DeploymentConfig.php]: Applied.

**Comment by Nuranto** [comment]
@ihor-sviziev have you succeed to reproduce the issue on your side ?

**Comment by ihor-sviziev** [comment]
@Nuranto I checked blame and it helped me to find place where it was introduced https://github.com/magento/magento2/pull/33803.
Please analyze it and try to understand why it's needed

**Comment by Nuranto** [comment]
Thank you ! It explains a lot indeed.
I'll work on something that could fix both issues.

**Comment by Nuranto** [comment]
I followed the steps described in https://github.com/magento/magento2/issues/33802 (https://github.com/magento/magento2/pull/33803) and I can't reproduce the issue described.

@sergeynezbritskiy Maybe you want to have a look at this PR. Even if this PR brings back the issue described in #33803, I think a proper fix would be to add a $deploymentConfig->resetData() where it's needed rather than reload data in case of null value.

**Comment by Nuranto** [comment]
@magento run Unit Tests

**Comment by Nuranto** [comment]
@magento run all tests

**Comment by Nuranto** [comment]
@magento run Database Compare, Functional Tests B2B, Functional Tests EE, Unit Tests, WebAPI Tests, Integration Tests

**Comment by jonathanribas** [comment]
Thanks @Nuranto for pointing this out!

We are live on Adobe Commerce 2.4.6 for a few days on a small market and noticed a CPU usage increase due to changes introduced on https://github.com/magento/magento2/issues/33802.
We were struggling to find out it was coming from this one... So we have now create a patch to run `2.4.5` DeploymentConfig version.

@ihor-sviziev you guys really need to run performance tests between Adobe Commerce releases, you should have pointed this out, impact was huge on 8th June:
Screenshot 2023-06-12 at 22 38 35

**Comment by aelmizeb** [comment]
@ihor-sviziev 
Can you please create an internal ticket to resolve the problem ? It will take eternity to see something to fix this big problem, knowing that there are already people under 2.4.6

**Comment by jonathanribas** [comment]
I completely agree with @aelmizeb, we should have at least a Quality Patch to fix this issue!

**Comment by ihor-sviziev** [comment]
@aelmizeb @jonathanribas I'm not an Adobe employee, can't do so :( 
@engcom-Hotel maybe you can raise it?

**Comment by Nuranto** [comment]
I think we should probably merge the fix asap for all impacted production environments, and then rework  [#33802](https://github.com/magento/magento2/issues/33802) later, with lower priority, since 33802 only concerns testing purpose.

(I gave up trying to validate this PR. I gave up on all my PRs actually, since it's very often failing for no reason, and then succeed after many relaunch... I'm willing to help, but this testing phase is a hell...)

**Comment by ihor-sviziev** [comment]
@Nuranto, could you please check if the change mentioned by @hostep  fixes the issue?

**Comment by Nuranto** [comment]
Maybe @alzota or @o-iegorov can tell us more about the goal of these commits ?

**Comment by jonathanribas** [comment]
Adobe Commerce support sent me this patch, it should be deployed soon for Adobe Commerce Cloud customers, hope we will get it from the official Quality Patches too...
[ACSD-51892_2.4.6.patch](https://github.com/magento/magento2/files/11755736/ACSD-51892_2.4.6.patch)

**Comment by ihor-sviziev** [comment]
@jonathanribas I see now that it has exactly the same changes as in https://github.com/magento/magento2/commit/6bc82de1deb09ce7a4a9d3b492c14363b338f128, so looks like the issue is already fixed in the 2.4-develop branch.

@Nuranto could you please confirm that?

**Comment by engcom-Hotel** [comment]
Hello @ihor-sviziev,

We will take this PR forward. So in that case I think we do not need any internal ticket.


Thanks

**Comment by ihor-sviziev** [comment]
Hi @Nuranto,
Thank you for confirmation! I'm closing this Pull Request since no actions are needed.

**Pull request closed without merging**